src/action-providers.ts

```diff
@@ -21,9 +21,73 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+    t.traverseAST(ast, {
+      enter: (path: t.NodePath<any>) => {
+        this.refactorings.forEach(
+          this.checkCanPerform(
+            path,
+            selection,
+            (
+              matchedPath: t.NodePath<any>,
+              refactoring: RefactoringWithActionProvider
+            ) => {
+              if (refactoring.actionProvider.canPerformRefactoringMutator) {
+                applicableRefactorings.push(
+                  refactoring.actionProvider.canPerformRefactoringMutator(
+                    matchedPath,
+                    refactoring
+                  )
+                );
+              } else {
+                applicableRefactorings.push(refactoring);
```

Comment:
I guess this is an intermediate step, until all refactorings are migrated to use the new pattern?

If so, I'd left a comment to explain "why" in the code (that could be surprising). But I think it's a good approach, so we can progressively ship this. I wouldn't like we have a long-living branch since it potentially conflict with other changes we could do.

Reply:
Actually, no, that's not what I thought. Some refactorings change how the action provider is displayed. Such refactorings can implement canPerformRefactoringMutator which mutates the refactoring object (which contains the actionProvider object), thus is able to mutate the action provider message. Refactorings that do not implement this function will just be added without any mutation happening.

Reply:
But I agree... It is not very easy to understand. Maybe, just maybe, refactorings could be a class that extends an abstract refactoring class which implements a no-op mutator. But usually, I don't like to use inheritance to obscure a simple condition. What do you think?